core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/RemoveRedundantCaseWhenLiteral.java

```diff
@@ -0,0 +1,149 @@
+import static io.trino.sql.tree.BooleanLiteral.FALSE_LITERAL;
+import static io.trino.sql.tree.ComparisonExpression.Operator.EQUAL;
+import static io.trino.sql.tree.ComparisonExpression.Operator.NOT_EQUAL;
+import static io.trino.sql.tree.LogicalExpression.Operator.OR;
+
+public class RemoveRedundantCaseWhenLiteral
+        implements Rule<FilterNode>
+{
+    private final PlannerContext plannerContext;
+    private final TypeAnalyzer typeAnalyzer;
+
+    public RemoveRedundantCaseWhenLiteral(PlannerContext plannerContext, TypeAnalyzer typeAnalyzer)
+    {
+        this.plannerContext = plannerContext;
+        this.typeAnalyzer = typeAnalyzer;
+    }
+
+    @Override
+    public Pattern<FilterNode> getPattern()
+    {
+        return filter();
+    }
+
+    @Override
+    public Result apply(FilterNode node, Captures captures, Context context)
+    {
+        Expression rewritten = ExpressionTreeRewriter.rewriteWith(new Visitor(plannerContext, context.getSession(), context.getSymbolAllocator(), typeAnalyzer), node.getPredicate());
+        if (node.getPredicate().equals(rewritten)) {
+            return Result.empty();
+        }
+
+        return Result.ofPlanNode(new FilterNode(node.getId(), node.getSource(), rewritten));
+    }
+
+    private static class Visitor
+            extends ExpressionRewriter<Void>
+    {
+        private final LiteralInterpreter literalInterpreter;
+        private final Session session;
+        private final SymbolAllocator symbolAllocator;
+        private final TypeAnalyzer typeAnalyzer;
+
+        public Visitor(PlannerContext plannerContext, Session session, SymbolAllocator symbolAllocator, TypeAnalyzer typeAnalyzer)
+        {
+            this.literalInterpreter = new LiteralInterpreter(plannerContext, session);
+            this.session = session;
+            this.symbolAllocator = symbolAllocator;
+            this.typeAnalyzer = typeAnalyzer;
+        }
+
+        @Override
+        public Expression rewriteComparisonExpression(ComparisonExpression node, Void context, ExpressionTreeRewriter<Void> treeRewriter)
+        {
+            if ((node.getOperator() == EQUAL || node.getOperator() == NOT_EQUAL)) {
+                if (node.getLeft() instanceof SearchedCaseExpression left &&
+                        node.getRight() instanceof Literal right &&
+                        left.getWhenClauses().stream().map(WhenClause::getResult).allMatch(Literal.class::isInstance) &&
+                        left.getDefaultValue().map(Literal.class::isInstance).orElse(false)) {
+                    Map<NodeRef<Expression>, Type> expressionTypes = typeAnalyzer.getTypes(session, symbolAllocator.getTypes(), node);
+                    Object rightValue = literalInterpreter.evaluate(right, expressionTypes.get(NodeRef.of(right)));
+
+                    Predicate<Expression> matcher = x -> (node.getOperator() == EQUAL) == rightValue.equals(literalInterpreter.evaluate(x, expressionTypes.get(NodeRef.of(x))));
```

Comment:
this comparison is hard to follow (a == b) == c. Can we rewrite it to make it more readable? I.e. extract it to a separate method and cover both EQUAL and NOT_EQUAL cases separately?

Reply:
Created another method to handle EQUAL and NOT_EQUAL separately.